In [1]:
pip install ucimlrepo

In [50]:
from ucimlrepo import fetch_ucirepo 
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression , ElasticNet, LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,MinMaxScaler,PolynomialFeatures
from sklearn.metrics import r2_score, mean_squared_error,confusion_matrix,accuracy_score
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import make_column_transformer
from sklearn.compose import make_column_selector
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import KFold , cross_val_score
from sklearn.impute import SimpleImputer

In [36]:
# fetch dataset 
breast_cancer_wisconsin_original = fetch_ucirepo(id=15) 
  
# data (as pandas dataframes) 
X = breast_cancer_wisconsin_original.data.features 
y = breast_cancer_wisconsin_original.data.targets 

In [37]:
X.columns

Index(['Clump_thickness', 'Uniformity_of_cell_size',
       'Uniformity_of_cell_shape', 'Marginal_adhesion',
       'Single_epithelial_cell_size', 'Bare_nuclei', 'Bland_chromatin',
       'Normal_nucleoli', 'Mitoses'],
      dtype='object')

In [38]:
y['Class'].unique()

array([2, 4], dtype=int64)

In [39]:
X.isnull().sum().sum()

16

In [40]:
X.drop('Bare_nuclei',axis=1,inplace=True)

C:\Users\DAI.STUDENTSDC\AppData\Local\Temp\ipykernel_17956\3673848491.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X.drop('Bare_nuclei',axis=1,inplace=True)


In [41]:
y['Class'].value_counts()

Class
2    458
4    241
Name: count, dtype: int64

In [42]:
y['Class'] = np.where(y['Class']==2,0,1)


C:\Users\DAI.STUDENTSDC\AppData\Local\Temp\ipykernel_17956\2243808073.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  y['Class'] = np.where(y['Class']==2,0,1)


In [43]:
y['Class'].value_counts()

Class
0    458
1    241
Name: count, dtype: int64

In [44]:
X_train , X_test , y_train , y_test = train_test_split(X,y,random_state=24,test_size=0.3)

In [45]:
lr = LogisticRegression()
lr.fit(X_train,y_train)
y_pred = lr.predict(X_test)
print(y_test['Class'].values)
print(y_pred)

[0 1 1 1 1 0 1 0 0 0 1 1 1 0 0 0 0 1 1 1 0 0 1 0 0 1 0 0 0 0 1 1 0 0 0 0 0
 1 0 0 0 0 0 0 1 0 1 1 1 0 0 0 1 0 0 0 0 1 0 0 1 0 1 0 0 1 1 0 0 1 0 0 0 0
 0 0 0 1 0 0 0 0 0 1 0 0 1 0 0 0 0 1 1 1 0 0 0 1 0 0 1 0 1 0 1 1 0 1 1 1 1
 0 0 0 0 1 0 0 1 0 1 0 0 1 0 0 0 0 0 0 0 0 0 1 1 0 1 0 1 0 1 0 1 0 0 0 0 1
 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 1 0 0 0 0 0 0 1 0 1 0
 0 1 1 1 0 1 0 0 1 1 0 1 1 0 0 1 0 1 0 0 0 1 0 0 1]
[0 1 1 1 1 0 1 0 0 0 1 1 1 0 0 0 0 1 1 1 0 0 1 0 0 1 0 0 0 0 1 1 0 0 0 0 0
 1 0 0 0 0 0 0 1 0 1 1 1 0 0 0 1 0 0 0 0 1 0 0 1 0 1 0 0 1 1 0 0 1 0 0 0 0
 0 0 0 1 0 0 0 0 0 1 0 0 1 0 0 0 0 1 1 1 0 0 0 1 0 0 1 0 0 0 1 1 0 1 1 1 1
 0 0 0 0 1 0 0 1 0 1 0 0 1 0 0 0 0 0 0 0 0 0 1 1 0 1 0 1 0 1 0 0 0 0 0 0 1
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 1 0 0 0 0 0 0 1 0 1 0
 0 1 1 1 0 1 1 0 1 1 0 1 1 0 0 1 0 1 0 1 0 0 0 0 1]


C:\Users\DAI.STUDENTSDC\AppData\Local\Anaconda3\Lib\site-packages\sklearn\utils\validation.py:1300: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [52]:
lr = LogisticRegression()
lr.fit(X_train,y_train)
y_pred = lr.predict(X_test)
pd.crosstab(y_test['Class'].values,y_pred)   # It is a confusion matrix (crosstab gives confusion matrix)

C:\Users\DAI.STUDENTSDC\AppData\Local\Anaconda3\Lib\site-packages\sklearn\utils\validation.py:1300: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


col_0    0   1
row_0         
0      135   3
1        4  68

In [53]:
# Accuracy score cal 
(135+68)/(135+3+4+68)

0.9666666666666667

In [54]:
print(confusion_matrix(y_test['Class'],y_pred))

[[135   3]
 [  4  68]]


In [55]:
accuracy_score(y_test['Class'],y_pred)

0.9666666666666667

In [56]:
y['Class'].value_counts(normalize=True)

Class
0    0.655222
1    0.344778
Name: proportion, dtype: float64

In [57]:
y['Class'].value_counts()

Class
0    458
1    241
Name: count, dtype: int64

##### Naive Rule / Baseline Model


In [60]:
y_pred = np.zeros(210)
accuracy_score(y_test['Class'],y_pred)

0.6571428571428571

##### Any ML Model score should be better than score of baseline model.
##### Here baseline (naive) model score = 0.6571428571428571 & ML model score = 0.9666666666666667.
##### Hence we can say that this ML model can be a relevant model.